libp2p/utils/semaphore.nim

```diff
@@ -65,7 +76,8 @@ proc release*(s: AsyncSemaphore) =
                             queue = s.queue.len
 
     if s.queue.len > 0:
-      var fut = s.queue.pop()
+      var fut = s.queue[0]
+      s.queue.del(0)
```

Comment:
`del` reorders the callbacks, you want `delete` here. the ordering test is too short to detect it - @cheatfate pointed out that the reordering might lead to starvation of some acquires if something else keeps getting ahead in the queue.

Reply:
That's only the case if deleting from the middle of the sequence? I would expect taking from the tip should keep the same order?

Reply:
it takes the last item and puts it first, because you're here deleting the first item. effectively, you're doing `pop` except for the first callback.

Reply:
heh, you're right:

```nim
proc del*[T](x: var seq[T], i: Natural) {.noSideEffect.} =
  ## Deletes the item at index `i` by putting ``x[high(x)]`` into position `i`.
  ##
  ## This is an `O(1)` operation.
  ##
  ## See also:
  ## * `delete <#delete,seq[T][T],Natural>`_ for preserving the order
  ##
  ## .. code-block:: Nim
  ##  var i = @[1, 2, 3, 4, 5]
  ##  i.del(2) # => @[1, 2, 5, 4]
  let xl = x.len - 1
  movingCopy(x[i], x[xl])
  setLen(x, xl)
```